### PR TITLE
[WIP] Groups and permissions

### PR DESCRIPTION
**Pull request by ujjwal-raizada**

### CODE REVIEW

File: backend/research/permissions.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+class IsStudent(permissions.BasePermission):
+    """
+    Custom permission to only allow student to access.
+    """
+
+    def has_permission(self, request, view):
+        if isinstance(request.user, Student):
+            return True
+        else:
+            return False
```
Comment (iamkroot) [inline backend/research/permissions.py]: This file should be moved to the users app, considering the permissions are common to all

File: backend/research/permissions.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+from research.models import ResearchScholar
+
+
+class IsHoD(permissions.BasePermission):
+    """
+    Custom permission to only allow HoD to access.
+    """
+
+    def has_permission(self, request, view):
+        if request.user.groups.filter(name='hod').exists():
```
Comment (iamkroot) [inline backend/research/permissions.py]: Should be shortened to `return request.user.groups.filter(name='hod').exists()`

File: backend/research/views.py [hunk 3]
```diff
@@ -19,12 +20,12 @@ class ProjectView(viewsets.ModelViewSet):
 
     def get_queryset(self):
         # TODO: HOD should be able to see all
-        return Project.objects.filter(author__id=self.request.user)
+        return Project.objects.filter(authors__email=self.request.user.email)
```
Comment (iamkroot) [inline backend/research/views.py]: no need to make explicit email field. If we specify `authors=user`, it will internally get converted to the pk

**Review by iamkroot — CHANGES_REQUESTED** [review]

File: backend/research/permissions.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            return False
+
+
+class IsFaculty(permissions.BasePermission):
+    """
+    Custom permission to only allow faculty to access.
+    """
+
+    def has_permission(self, request, view):
+        if isinstance(request.user, Faculty):
```
Comment (iamkroot) [inline backend/research/permissions.py]: `isinstance` might not be the best way to go, as we are doing a very redundant check here. Better to use [Model Permissions](https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#custom-permissions) with `request.user.has_perm` and assign these permissions at the Model level. See [DRF Model Perms](https://www.django-rest-framework.org/api-guide/permissions/#djangomodelpermissions).

**Comment by iamkroot** [comment]
Run a rebase from master, squash unnecessary commits

**Comment by ujjwal-raizada** [comment]
@iamkroot rebased and squashed

**Review by iamkroot — APPROVED** [review]

**Pull request merged**